### ISSUE TITLE
Split up and merge NWT

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#546]
This is getting messy: lets collect everything in a GitHub issue so that we can easily link to PRs, branches, and commits.

For those of us joining us now: the NWT branch contains a ton of new examples (and a fair number of new features) but the history is difficult to understand because there are merge commits, application of tools like clang-format, etc. @amneetb, @boyceg, and myself are trying to split up the branch into more reasonable chunks.

Please edit this list until we have a comprehensive list of *all* the stuff in `amneetb/numerical-wave-tank` that we want to keep.
## commits with new functionality
- [ ] whatever commit added the stuff we need for 74f7494656.
- [ ] commit a11a5fa265 changes the pressure extraction algorithm for multiphase flows to second order.
- [ ] commits b142914b19, dcadb2845b, and 729f222d4c added a new Fortran routine for sign sweeping.
- [x] commit a710b9a582 added Brinkman penalization code (see #625).
- [x] commit c188e2116d enables translation (without rotation) in `ConstraintIB{Method,Kinematics}` (see #567)
- [ ] commit b300bc90ae adds some postprocessing to `ConstraintIBMethod`
- [x] commit 5c171743ef adds `FESurfaceDistanceEvaluator` (see also 16ef35b9f6) (see #580)
- [ ] commits 53cd8f1708 and a95c6b3527 fix some bugs in `IBLevelSet/ex0`
- [x] ~~commit 14addfd130 makes `regridProjection` public in some classes.~~ See the discussion in #527 for more information on why this method should remain `protected.

## commits adding new examples
- [x] commit 74f7494656 : three new multiphase flow examples (ex7, ex8, and ex9) (requires `ConstraintIBMethod::getFullLagrangeMultiplierForce` and `ConstraintIBMethod::getInterpolatedLagrangianDensity`) (we also need to update authorship here; that ended up in another commit) (see also bb6ae337367) (tracked by #547)
- [x] commit a9a3c3053a334 : fix an issue with regrid intervals in two-phase examples (this is independent of any other changes AFAICT) (tracked by #550)
- [x] commit 00bf3a616938 : four new multiphase flow examples (ex10, ex11, ex12, ex13) (requires `ConstraintIBMethod::getFullLagrangeMultiplierForce` and `ConstraintIBMethod::getInterpolatedLagrangianDensity`) (see also 5641adb47998, 25d881701ed, bb6ae337367322, and 76620bcb2050) (tracked by #548)

## commits with examples and functionality
- [ ] commits 8de1786c81, 6a7110e963, 76620bcb20, bb6ae33736, 7a7ff35514, b4b504f6e6, 3ac309067c, 56e578a76b, 7e4ab5122e, b473b8d087, 8631d6d47a, 300dca0434, 25d881701e, 15cc02339c, a3a8599f3b, 5641adb479, d54525a477, 7144e2e121, 3940f8ccb3, e1491a2dd1, 89c6da60c7 and 8a88364911 added both the wave tank files and examples. Most commits change both the implementation and the examples: we will need to take these apart.
- [x] ~~commit 8a88364911 changes the INSVC integrators by adding a damping zone in addition to changing some examples.~~ update: this didn't work and should be removed.
- [ ] commits 7144e2e121, baee7b62f6 and 737d1a39aa add `IB{LevelSet,Interpolant}Method` and `IBInterpolantHierarchyIntegrator` and also changes a wave tank example. The first commit also changes `HierarchyIntegrator::atRegridPointSpecialized`. 737d1a39aa also changes `IBHydrodynamicSurfaceForceEvaluator`. (see #607)
- [ ] commits 16ef35b9f6, f674f25c45af edits `FESurfaceDistanceEvaluator` and uses it in some IBLevelSet examples
- [x] commits 1af1de62d4, 5bff8ec1c8, and 1a020fc20d add the Brinkman code. 1af1de62d4 also changes *a lot* of other classes (`CIBMethod`, `FESurfaceDistanceEvaluator`, `IBFEDirectForcingKinematics`, etc) (see #625 and #626)
- [ ] ff21fb7607 and 1af1de62d4 implement `IBLevelSet/ex0` (and also change the Brinkman code)

The examples require additional changes made on NWT so I suggest that we either combine them with the commits containing relevant functionality or merge them afterwards.

In summary, NWT consists of multiple major changesets:
1. the new wave tank code
2. the Brinkman code (and corresponding changes to the INSVC code)
3. some changes to the `CIB` code
4. FESurfaceDistanceEvaluator
5. new level set code
6. damping code in INSVC (possibly related to Brinkman?)
7. multiphase examples
8. wave tank examples

Note: one can see all changes made on this branch (excluding interleaved changes on `master` that show up due to repeated merges) by running `git log --no-merges amneetb/numerical_wave_tank ^master`.

**Comment by boyceg** [comment]
> The examples require additional changes made on NWT so I suggest that we either combine them with the commits containing relevant functionality or merge them afterwards.

Suggest to make the examples a separate commit or PR that follows the additional functionality.

**Comment by amneetb** [comment]
@drwells : Commit bb6ae33 removed `ConstraintIBMethod::getFullLagrangeMultiplierForce` and `ConstraintIBMethod::getInterpolatedLagrangianDensity` from the library and examples. Can you check this commit to see if it resolves one of issues.

**Comment by drwells** [comment]
I believe that I have now exhaustively listed all changes in the NWT branch. There are 18 bullet points in the above list: I strongly suggest that we write at least 18 PRs to get all of the functionality in. I believe @amneetb should be in charge of this since he understands the requirements and interdependencies best.

**Comment by amneetb** [comment]
> This is getting messy: lets collect everything in a GitHub issue so that we can easily link to PRs, branches, and commits.
> 
> For those of us joining us now: the NWT branch contains a ton of new examples (and a fair number of new features) but the history is difficult to understand because there are merge commits, application of tools like clang-format, etc. @amneetb, @boyceg, and myself are trying to split up the branch into more reasonable chunks.
> 
> Please edit this list until we have a comprehensive list of _all_ the stuff in `amneetb/numerical-wave-tank` that we want to keep.
> 
> ## commits with new functionality
> * [ ]  whatever commit added the stuff we need for [74f7494](https://github.com/IBAMR/IBAMR/commit/74f7494656e14da52b64a863d7e57e324d0ecb5b).
> * [ ]  commit [a11a5fa](https://github.com/IBAMR/IBAMR/commit/a11a5fa2656633f9072dd6f6f6440e2cf838ab93) changes the pressure extraction algorithm for multiphase flows to second order.
> * [ ]  commits [b142914](https://github.com/IBAMR/IBAMR/commit/b142914b190f8e1e796f459e4461e87fb05a53d1), [dcadb28](https://github.com/IBAMR/IBAMR/commit/dcadb2845b0712a10299983eac588847f2adb829), and [729f222](https://github.com/IBAMR/IBAMR/commit/729f222d4c4c40c45540f22c3d9e0b052169b9ce) added a new Fortran routine for sign sweeping.
> * [ ]  commit [a710b9a](https://github.com/IBAMR/IBAMR/commit/a710b9a582504a91d13674c807ac39bf8367cab0) added Brinkman penalization code.
> * [ ]  commit [c188e21](https://github.com/IBAMR/IBAMR/commit/c188e2116d2ed7fc05bec29412573dfd0c775483) enables translation (without rotation) in `ConstraintIB{Method,Kinematics}`
> * [ ]  commit [b300bc9](https://github.com/IBAMR/IBAMR/commit/b300bc90aefb2a608cfddc7cfa59d458ac6b286f) adds some postprocessing to `ConstraintIBMethod`
> * [ ]  commit [5c17174](https://github.com/IBAMR/IBAMR/commit/5c171743ef8e4187313054b6c76c01532e3b0cb6) adds `FESurfaceDistanceEvaluator` (see also [16ef35b](https://github.com/IBAMR/IBAMR/commit/16ef35b9f69943f0a63b9228552cf313a8a826a0))
> * [ ]  commits [53cd8f1](https://github.com/IBAMR/IBAMR/commit/53cd8f170853e90ebb9fbc7d10e69fe69ad7c642) and [a95c6b3](https://github.com/IBAMR/IBAMR/commit/a95c6b3527c301c300363966726e41ce3fd54a66) fix some bugs in `IBLevelSet/ex0`
> * [ ]  commit [14addfd](https://github.com/IBAMR/IBAMR/commit/14addfd13027fb0897aced0357e74ee2f87994bf) makes `regridProjection` public in some classes. See the discussion in #527 for more information.
> 
> ## commits adding new examples
> * [ ]  commit [74f7494](https://github.com/IBAMR/IBAMR/commit/74f7494656e14da52b64a863d7e57e324d0ecb5b) : three new multiphase flow examples (ex7, ex8, and ex9) (requires `ConstraintIBMethod::getFullLagrangeMultiplierForce` and `ConstraintIBMethod::getInterpolatedLagrangianDensity`) (we also need to update authorship here; that ended up in another commit) (see also [bb6ae33](https://github.com/IBAMR/IBAMR/commit/bb6ae33736732233a0f10e50a54d5fa9a51ac37a)) (tracked by #547)
> * [ ]  commit [a9a3c30](https://github.com/IBAMR/IBAMR/commit/a9a3c3053a334d3f4dd92882ebb0ffe8c6cd519e) : fix an issue with regrid intervals in two-phase examples (this is independent of any other changes AFAICT) (tracked by #550)
> * [ ]  commit [00bf3a6](https://github.com/IBAMR/IBAMR/commit/00bf3a616938e00c7552cf9bdcf2d08380ee4ea5) : four new multiphase flow examples (ex10, ex11, ex12, ex13) (requires `ConstraintIBMethod::getFullLagrangeMultiplierForce` and `ConstraintIBMethod::getInterpolatedLagrangianDensity`) (see also [5641adb](https://github.com/IBAMR/IBAMR/commit/5641adb479982426867c1b2fca130e38cb176ba5), [25d8817](https://github.com/IBAMR/IBAMR/commit/25d881701ed3fb1e57890232dd4b10844fa37997), [bb6ae33](https://github.com/IBAMR/IBAMR/commit/bb6ae33736732233a0f10e50a54d5fa9a51ac37a), and [76620bc](https://github.com/IBAMR/IBAMR/commit/76620bcb2050fa270adf1baf316f34bc3cbf4430)) (tracked by #548)
> 
> ## commits with examples and functionality
> * [ ]  commits [8de1786](https://github.com/IBAMR/IBAMR/commit/8de1786c8190574944839309846c0ebeada66a27), [6a7110e](https://github.com/IBAMR/IBAMR/commit/6a7110e96352195c55f3cd974a34aa8127d9841a), [76620bc](https://github.com/IBAMR/IBAMR/commit/76620bcb2050fa270adf1baf316f34bc3cbf4430), [bb6ae33](https://github.com/IBAMR/IBAMR/commit/bb6ae33736732233a0f10e50a54d5fa9a51ac37a), [7a7ff35](https://github.com/IBAMR/IBAMR/commit/7a7ff35514b93626fd54438a2aa257091b92cfbf), [b4b504f](https://github.com/IBAMR/IBAMR/commit/b4b504f6e6aec78cb07fabd44eb0d3dbea0eea0f), [3ac3090](https://github.com/IBAMR/IBAMR/commit/3ac309067c7c6e2c4cedd74882d4faa687f10bc9), [56e578a](https://github.com/IBAMR/IBAMR/commit/56e578a76bda6eaac5fee87a0cc5d14d1fc87467), [7e4ab51](https://github.com/IBAMR/IBAMR/commit/7e4ab5122e3dc810ddc2ff4af6191c454d49b258), [b473b8d](https://github.com/IBAMR/IBAMR/commit/b473b8d087ffa48972ac27a7fd46789c919a58d1), [8631d6d](https://github.com/IBAMR/IBAMR/commit/8631d6d47ae7f88a4d5171431e5a3b091aa50baa), [300dca0](https://github.com/IBAMR/IBAMR/commit/300dca0434159089a2ea1955e5298ce5d3d18c41), [25d8817](https://github.com/IBAMR/IBAMR/commit/25d881701ed3fb1e57890232dd4b10844fa37997), [15cc023](https://github.com/IBAMR/IBAMR/commit/15cc02339c1bcc1334ad5ba9dbab1bb94fd012c1), [a3a8599](https://github.com/IBAMR/IBAMR/commit/a3a8599f3b91756e720f318f4d076c37629c76bd), [5641adb](https://github.com/IBAMR/IBAMR/commit/5641adb479982426867c1b2fca130e38cb176ba5), [d54525a](https://github.com/IBAMR/IBAMR/commit/d54525a4779a587c22359a05da5509d8c519f69a), [7144e2e](https://github.com/IBAMR/IBAMR/commit/7144e2e1216afb970e0b58aa4a7fbbaa747730e5), [3940f8c](https://github.com/IBAMR/IBAMR/commit/3940f8ccb34de76f161887914c6ba93c414fbc58), [e1491a2](https://github.com/IBAMR/IBAMR/commit/e1491a2dd153fcbe76e6d4ccc5cb73fb7ca37cb6), [89c6da6](https://github.com/IBAMR/IBAMR/commit/89c6da60c7e555f1e6fcd37227ebf8eed31d74bd) and [8a88364](https://github.com/IBAMR/IBAMR/commit/8a88364911ecb40e017f5f72f67f59ef9be29e66) added both the wave tank files and examples. Most commits change both the implementation and the examples: we will need to take these apart.
> * [ ]  commit [8a88364](https://github.com/IBAMR/IBAMR/commit/8a88364911ecb40e017f5f72f67f59ef9be29e66) changes the INSVC integrators by adding a damping zone in addition to changing some examples.

@drwells : This functionality of implicit damping zone has been removed from the branch. It was not working well.

> * [ ]  commits [7144e2e](https://github.com/IBAMR/IBAMR/commit/7144e2e1216afb970e0b58aa4a7fbbaa747730e5), [baee7b6](https://github.com/IBAMR/IBAMR/commit/baee7b62f6282656aaf0a9825180945c52d84a54) and [737d1a3](https://github.com/IBAMR/IBAMR/commit/737d1a39aae52840336ea0efbd815d1ef3f628dd) add `IB{LevelSet,Interpolant}Method` and `IBInterpolantHierarchyIntegrator` and also changes a wave tank example. The first commit also changes `HierarchyIntegrator::atRegridPointSpecialized`. [737d1a3](https://github.com/IBAMR/IBAMR/commit/737d1a39aae52840336ea0efbd815d1ef3f628dd) also changes `IBHydrodynamicSurfaceForceEvaluator`.
> * [ ]  commits [16ef35b](https://github.com/IBAMR/IBAMR/commit/16ef35b9f69943f0a63b9228552cf313a8a826a0), [f674f25](https://github.com/IBAMR/IBAMR/commit/f674f25c45afd6f93c7c133cdd8c700e85236303) edits `FESurfaceDistanceEvaluator` and uses it in some IBLevelSet examples
> * [ ]  commits [1af1de6](https://github.com/IBAMR/IBAMR/commit/1af1de62d41f3ae9f3b9ff38c2b6d852ad4b379e), [5bff8ec](https://github.com/IBAMR/IBAMR/commit/5bff8ec1c8940f6d0a4f726ce74a36d4b2845ff2), and [1a020fc](https://github.com/IBAMR/IBAMR/commit/1a020fc20df3067a4f4f49e67f79f3d108ca4026) add the Brinkman code. [1af1de6](https://github.com/IBAMR/IBAMR/commit/1af1de62d41f3ae9f3b9ff38c2b6d852ad4b379e) also changes _a lot_ of other classes (`CIBMethod`, `FESurfaceDistanceEvaluator`, `IBFEDirectForcingKinematics`, etc)
> * [ ]  [ff21fb7](https://github.com/IBAMR/IBAMR/commit/ff21fb7607bc14e045831b6a3bf6aeb40133e876) changes the Brinkman code and also `IBLevelSet/ex0`
> 
> The examples require additional changes made on NWT so I suggest that we either combine them with the commits containing relevant functionality or merge them afterwards.
> 
> In summary, NWT consists of multiple major changesets:
> 
> 1. the new wave tank code
> 2. the Brinkman code (and corresponding changes to the INSVC code)
> 3. some changes to the `CIB` code
> 4. FESurfaceDistanceEvaluator
> 5. new level set code
> 6. damping code in INSVC (possibly related to Brinkman?)
> 7. multiphase examples
> 8. wave tank examples
> 
> Note: one can see all changes made on this branch (excluding interleaved changes on `master` that show up due to repeated merges) by running `git log --no-merges amneetb/numerical_wave_tank ^master`.

**Comment by amneetb** [comment]
@drwells : To be specific the implicit damping zone was removed in this commit 1af1de62d41f3ae9f3b9ff38c2b6d852ad4b379e

(This code however served as a testbed for more general Brinkmann penalization method)

**Comment by amneetb** [comment]
> I believe that I have now exhaustively listed all changes in the NWT branch. There are 18 bullet points in the above list: I strongly suggest that we write at least 18 PRs to get all of the functionality in. I believe @amneetb should be in charge of this since he understands the requirements and interdependencies best.

Me in charge of 18 PRs? I think it is a bad idea. I can work on your comments in these PRs, but I don't have the bandwidth to break and manage them. Sorry.

**Comment by drwells** [comment]
We will have to find someway of chopping them up: I know very little about the subject material for most of these changes so it will take me some time to figure out what chunks are relevant, what changes go with what examples, etc. I look for some other simple ones once we get the first three dealt with but we really do need to split up the work on this  branch.

**Comment by amneetb** [comment]
I am not good with Git, but I am happy to work on any PR you create.

(Like I have mentioned before, there is not much to this PR. There are bunch of new files and new examples. What I have in my head is quite simple, but with Git complexity it seems daunting. I leave it unto you to break the PR, rewrite history etc.)

**Comment by amneetb** [comment]
@drwells : In the interest of time and to make everyone's life easier, I suggest that we follow Boyce's suggestion of creating a diff between NWT and master and making a new branch of it.

**Comment by amneetb** [comment]
At this point, commit authorship, history etc., is just not worth preserving. Let's get this done and move on.

**Comment by drwells** [comment]
The comment I wrote here an hour ago seems to have vanished. I will try and recreate what I wrote.

I agree that we shouldn't worry about preserving the history (though, of course, we need to write changelog entries so that @amneetb and @nishantnangia get credit for all this work!)

I talked with @boyceg about our strategy for this branch on Tuesday and we did discuss merging to get that branch up to date. However: my understanding of that conversation was that the *goal* of merging was that we could get #418 to the point where we could determine what has actually changed (excluding #547, #548, and #550) and inventory the changes on the branch. I was able to do this without a merge with some careful calls to `git-log` so I did not merge. @boyceg its possible that I misunderstood our conclusions: should I, regardless of the inventory above, go back and handle the merge?

Even if we get the merge conflicts sorted out there is still a substantial amount of work to do on this branch before we are done: I am hesitant to merge the branch as-is when the changes in #550 either expose or create segmentation faults in the library. Additionally, examples like `IBLevelSet/ex0` still assume that `regridProjection` is `public`, but we reached a consensus that that function should be `protected`, so they will not compile at the moment on master. It is not clear to me how we should fix this.

Furthermore: as noted by @boyceg in in https://github.com/IBAMR/IBAMR/pull/418#issuecomment-484369373: there are is a lot of application specific code in #418 that really should not be in the core library (e.g., the heart model code is not in the library itself). I agree with @boyceg that this really belongs somewhere else.

**Comment by boyceg** [comment]
> talked with @boyceg about our strategy for this branch on Tuesday and we did discuss merging to get that branch up to date. However: my understanding of that conversation was that the _goal_ of merging was that we could get #418 to the point where we could determine what has actually changed (excluding #547, #548, and #550) and inventory the changes on the branch. I was able to do this without a merge with some careful calls to `git-log` so I did not merge. @boyceg its possible that I misunderstood our conclusions: should I, regardless of the inventory above, go back and handle the merge?

I should be able to merge master into the current branch and resolve the conflicts later tonight.

**Comment by boyceg** [comment]
I just pushed a new branch (`amneetb/numerical_wave_tank-2 ` for now) that merges master into `amneetb/numerical_wave_tank`. If this looks good, we can cherry-pick onto `amneetb/numerical_wave_tank`.

I also can setup the diff branch, but I think we should merge the examples to master first, to minimize the number of changes.

**Comment by drwells** [comment]
Yes. Lets continue merging examples before proceeding to the `diff` branch.

**Comment by amneetb** [comment]
Let's aim to finish this soon.

**Comment by drwells** [comment]
I am talking with @amneetb about this now (I have to write this down or else I will forget).

We should try to get the Brinkman code and the INSVCHierarchyIntegrator fixes next so that an incoming student can get going on some projects.

@amneetb also wants to get the different wave tank boundary condition code merged soon too.

**Comment by drwells** [comment]
@boyceg let me know if you want me to split off any more parts of this before you go to town on it.

**Comment by boyceg** [comment]
If you can split off any more today, feel free to do so. I will not start working on this until some time tomorrow. Today I am trying to finish up revisions to IBFESurfaceMethod.

**Comment by drwells** [comment]
I have spent a ton of time on this (at least 40-50 hours) over the last few months and we are only about halfway through. This branch still has a ton of new stuff in it that we need to get into IBAMR. I am moving it to 0.6 but I suspect we won't get everything done: its just too much if we want to stick to the tentative plan of finishing things up by the start of the semester.

Lets focus on #580 next.

**Comment by amneetb** [comment]
@boyceg : Would you be able to review it this month?

**Comment by drwells** [comment]
This has been superceded by #744, which collates the remaining changes into a few large commits.
